File: 98_validate_BST/validBST.cpp
#include <iostream>

struct TreeNode{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution{
public:
    
  bool isValidBST(TreeNode* root) {
        TreeNode* prev = NULL;
        return validate(root, prev);
    }
    
    /**
     * @brief   Use inorder traversal to check the validity of BST in O(n)
     * @details Traverse the BST in-order and check if it is in ascending order.
     *          To do this, instead of mainining the entire traversal, just compare if 
     *          previous element is less than the current element. Thus preserving only 
     *          the address of previous element.
     * @param root  Current element
     * @param prev  Previous element std::nullptr by default
     * @returns bool, true if Binary search sub-tree is valid
     * 
     */
    bool validate(TreeNode* node, TreeNode* &prev) {
        /* Inorder - left, right  root */
        /* Exit case */
        if (node == NULL) return true;
        
        /* Left */
        if (!validate(node->left, prev)) return false;
        
        /* Root - remember to save current node as prev for next iteration */
        if (prev != NULL && prev->val >= node->val) return false;
        prev = node;
        
        /* Right */
        return validate(node->right, prev);
    }
};

File: 98_validate_BST/validBST.py
class TreeNode(object):
    def __init__:(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):     
    def validate(self, node: TreeNode, prev) -> bool:
        """
        Inorder Traversal
        Exit condition:
        """
        if (node == None):
            return true
    
        # left sub tree
        if self.validate(node.left, prev) is False:
            return False

        # Root Node
        if prev and prev.val >= node.val:
            return False

        # Right node
        prev = node
        return self.validate(node.right, prev)

    def isValidBST(self, root: TreeNode) -> bool:
        """
        :type root: TreeNode
        :rtype: bool
        """
        prev = None
        return validate(self, root, prev)
        
    
